san/impl/derive_type.cpp

```diff
@@ -0,0 +1,256 @@
+void DeriveType::visit(IfStatement* node) {
+    node->condition->accept(*this);
+    // check condition is of type boolean
+    node->ifBody->accept(*this);
+    if (node->elseBody != nullptr) {
+        node->elseBody->accept(*this);
+    }
+}
+
+void DeriveType::visit(Expression*) {}
+
+void DeriveType::visit(UnaryExpression* node) {
+    node->operand->accept(*this);
+    node->type = node->operand->type;
+}
+
+void DeriveType::visit(BinaryExpression* node) {
+    //  set correct type
+    if (node->operation == lexer::TokenType::OpenBrack) {
+        // if operation is array access
+        // check that the first item is array -> not nec a[1][0]
+        m_searchArray = true;
+        node->operand1->accept(*this);
+        if (m_arrayInnerType == nullptr) {
+            error(node->operand1->begin,
+                  "invalid operation [] on the given type");
+        }
+        m_searchArray = false;
+        node->operand2->accept(*this);
+        if (!typeIsBase(node->operand2->type)) {
+            error(node->operand2->begin, "invalid type for array index");
+        }
+        node->type = m_arrayInnerType;
+    } else if (node->operation == lexer::TokenType::Dot) {
+        // if operation is dot ntation
+        // check that the first operand is a record -???
+        // take the type of key arg
+        m_searchFiled = true;
+        node->operand2->accept(*this);
+        m_searchFiled = false;
+        if (m_recordField == "") {
+            error(node->operand2->begin, "field name of record not found");
+        }
+        // search for the type of key arg
+        m_searchRecord = true;
+        node->operand1->accept(*this);
+        m_searchRecord = false;
+
+        if (m_recordInnerType == nullptr) {
+            error(node->operand1->begin,
+                  "invalid operation '.' on the given type");
+        }
+
+        node->type = m_recordInnerType;
+
+    } else if (node->operand1->type != nullptr &&
+               node->operand2->type != nullptr &&
+               node->operand1->type->getTypeKind() !=
+                   node->operand2->type->getTypeKind()) {
+        sPtr<ast::Type> type1 = node->operand1->type;
+        if (!typeIsBase(type1)) {
+            error(node->operand1->begin, "invalid type of expression");
+        }
+
+        sPtr<ast::Type> type2 = node->operand2->type;
+        if (!typeIsBase(type2)) {
+            error(node->operand2->begin, "invalid type of expression");
+        }
+
+        node->type = getGreaterType(type1, type2);
+    } else {
+        // types are equal
+        node->type = node->operand1->type;
+    }
+}
+
+void DeriveType::visit(Primary*) {}
+
+void DeriveType::visit(IntegerLiteral*) {}
+
+void DeriveType::visit(RealLiteral*) {}
+
+void DeriveType::visit(BooleanLiteral*) {}
+
+void DeriveType::visit(Identifier* node) {
+    node->variable->accept(*this);
+
+    if (m_searchFiled) {
+        m_recordField = node->name;
+    }
+    node->type = node->variable->type;
```

Comment:
`IdentifierResolver` already does this

Reply:
No. Some examples fail due to the fact, that Identifier don't have type.

Reply:
If it does not have a type, that should mean that it is not linked to a variable declaration because this is literally the job of the other visitor. This happened primarily in the case of record/array access (and should have probably been fixed there, not here). Anyways, when this is the case, it is highly probably that `node->variable` is null, which would make the first statement fail